### PR TITLE
Ft/openAfrica API integration

### LINKED ISSUE #411: @charterafrica/openafrica
Add fetching datasets from [openAfrica](https://openafrica.net/)

Detailed document [here](https://docs.google.com/document/d/14iqZy4cAXuav1uxk0fEEb5JTQiM2Oli_o_px6g8bh7M/edit?usp=sharing)

### PR DESCRIPTION
**Pull request by kelvinkipruto**
## Description

This PR adds the feature to fetch documents from [openAfrica](https://openafrica.net/).

Fixes #411

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: apps/charterafrica/src/pages/api/v1/data/datasets.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      const datasets = await fetchDatasets(q, {
+        fq: filterQuery,
+      });
+      res.status(200).json(datasets);
+    } catch (error) {
+      res.status(500).json({ error });
+    }
+  }
+
+  return res.json(404).json({ message: "NO ORGANIZATION PROVIDED" });
```
Comment (koechkevin) [inline apps/charterafrica/src/pages/api/v1/data/datasets.js:29]:
```suggestion
  return res.status(404).json({ message: "NO ORGANIZATION PROVIDED" });
```

**Review by koechkevin — CHANGES_REQUESTED** [review]

**Review by koechkevin — COMMENTED** [review]
Overall, the code appears to be well-structured and follows best practices for handling API requests and responses.

**Review by kilemensi — COMMENTED** [review]
👍🏽 

---

Haven't run it locally but I'm iffy about 2 things:
1. Why do we need the front-end to send the `organisation` via `v1/data/datasets`? I'm assuming we don't want this to change i.e. it should always point to the one set in the CMS.
2. Don't see anywhere we're handling pagination.

**Review by koechkevin — APPROVED** [review]

**Comment by kilemensi** [comment]
> 👍🏽
> 
> Haven't run it locally but I'm iffy about 2 things:
> 
>     1. Why do we need the front-end to send the `organisation` via `v1/data/datasets`? I'm assuming we don't want this to change i.e. it should always point to the one set in the CMS.
> 

Heita @kelvinkipruto .

Where are we addressing point 1 above? i.e. pull organisation id from the CMS and use it for searching. Was hoping to see it somewhere in `src/lib/openafrica`

**Comment by kilemensi** [comment]
@kelvinkipruto ?

**Comment by kelvinkipruto** [comment]
@kilemensi Sorry I was updating how it fetches `tags`
In my design approach `src/lib/openafrica` should only interact with `openAfrica API` and not with the CMS. Getting data from CMS should be handled similarly to how we do in `sourceAfrica`. I have this [file](https://github.com/CodeForAfrica/ui/blob/f0eccf2697a34a474e84a1542ecfcbac211671f4/apps/charterafrica/src/lib/data/common/processPageDatasets.js) to show how we can handle it.
The organization will be passed to the component as [props](https://github.com/CodeForAfrica/ui/blob/e64e9666ec419163ed871078c81071cebc91ad55/apps/charterafrica/src/lib/data/common/processPageDatasets.js#L19) which we can then use when making API calls to `v1/data/datasets`

**Comment by kelvinkipruto** [comment]
@kilemensi We [only fetch data](https://github.com/CodeForAfrica/ui/blob/e64e9666ec419163ed871078c81071cebc91ad55/apps/charterafrica/src/lib/data/common/processPageDatasets.js#L8-L9) if we have an `organization` provided in the CMS. The `organizationId` is required, so we're guaranteed to always have an organization to fetch data from. So yes, we can fetch data from any organization that is set up in the CMS.

**Comment by kilemensi** [comment]
> @kilemensi We [only fetch data](https://github.com/CodeForAfrica/ui/blob/e64e9666ec419163ed871078c81071cebc91ad55/apps/charterafrica/src/lib/data/common/processPageDatasets.js#L8-L9) if we have an `organization` provided in the CMS. The `organizationId` is required, so we're guaranteed to always have an organization to fetch data from. So yes, we can fetch data from any organization that is set up in the CMS.

@kelvinkipruto I mean the API accepts **any** organisation not necessarily the one set in the CMS.

**Comment by kelvinkipruto** [comment]
@kilemensi Yes, now I get you. The API will accept any organization passed to it.
Given that I have set the `datasets` as a page block, it would be difficult to fetch the `organizationId` from `src/lib/openafrica`. To solve it perhaps I could set datasets as a `collection`. This will make it possible to fetch the collection, hence the organization easily

**Comment by kelvinkipruto** [comment]
@kilemensi I have updated it.

File: apps/charterafrica/src/lib/data/rest/index.js [hunk 2]
```diff
@@ -53,6 +53,12 @@ export const api = {
   findPage,
 };
 
+export async function getOrganizationId() {
+  const data = await findGlobal("datasets");
+  const { organizationId } = data;
+  return organizationId;
+}
+
```
Comment (kilemensi) [inline apps/charterafrica/src/lib/data/rest/index.js:61]: Not sure this is necessary... both pages via lib/data and APIs have access to `payload` object and can hence fetch the organisation `id`/`slug` _**before**_ passing it to `openAfrica` lib.

The `openAfrica` lib shouldn't need to know anything about data/rest, etc.

File: apps/charterafrica/src/lib/data/common/processPageDatasets.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import fetchDatasets from "@/charterafrica/lib/openAfrica";
+
+async function processPageDatasets(page) {
+  const { blocks } = page;
+
```
Comment (kilemensi) [inline apps/charterafrica/src/lib/data/common/processPageDatasets.js:5]:
```suggestion
```

File: apps/charterafrica/src/lib/data/common/processPageDatasets.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      rows: 10,
+      start: 0,
+    });
+
+    blocks[datasetsIndex] = {
+      slug: "datasets",
+      datasets,
+    };
+  }
+
```
Comment (kilemensi) [inline apps/charterafrica/src/lib/data/common/processPageDatasets.js:19]:
```suggestion
```

**Review by kilemensi — COMMENTED** [review]
🚀

**Pull request merged**